### PR TITLE
Create get properties msg

### PR DESCRIPTION
**Pull request by pedropgusmao**

### CODE REVIEW

File: src/proto/flwr/proto/transport.proto [hunk 1]
```diff
@@ -69,7 +69,7 @@ message ClientMessage {
     float accuracy = 3 [ deprecated = true ];
     map<string, Scalar> metrics = 4;
   }
-
+  message GetInfo { map<string, Scalar> info = 1; }
```
Comment (danieljanes) [inline src/proto/flwr/proto/transport.proto:72]: `GetProperties`

File: src/py/flwr/server/client_proxy.py [hunk 2]
```diff
@@ -33,11 +36,15 @@ class ClientProxy(ABC):
 
     def __init__(self, cid: str):
         self.cid = cid
+        self.properties: Properties = {}
 
     @abstractmethod
     def get_parameters(self) -> ParametersRes:
         """Return the current local model parameters."""
 
+    def get_properties(self, ins: PropertiesIns) -> PropertiesRes:
```
Comment (danieljanes) [inline src/py/flwr/server/client_proxy.py:45]: This needs an `@abstractmethod`

File: src/py/flwr/simulation/ray_transport/ray_client_proxy.py [hunk 3]
```diff
@@ -46,6 +46,17 @@ def get_parameters(self) -> common.ParametersRes:
             res,
         )
 
+    def get_properties(self, ins: common.PropertiesIns) -> common.PropertiesRes:
+        """Returns client's properties."""
+        future_get_properties_res = launch_and_get_properties.options(
```
Comment (danieljanes) [inline src/py/flwr/simulation/ray_transport/ray_client_proxy.py:51]:
```suggestion
        future_properties_res = launch_and_get_properties.options(
```

File: src/proto/flwr/proto/transport.proto [hunk 4]
```diff
@@ -44,12 +44,13 @@ message ServerMessage {
     Parameters parameters = 1;
     map<string, Scalar> config = 2;
   }
-
+  message PropertiesIns { map<string, Scalar> properties = 1; }
```
Comment (danieljanes) [inline src/proto/flwr/proto/transport.proto:47]:
```suggestion
  message PropertiesIns { map<string, Scalar> config = 1; }
```

Comment (danieljanes) [reply src/proto/flwr/proto/transport.proto:47]: @pedropgusmao not that this renaming probably requires a few downstream renamings as well

**Review by danieljanes — CHANGES_REQUESTED** [review]

File: src/py/flwr/client/numpy_client.py [hunk 5]
```diff
@@ -86,6 +89,17 @@ def get_parameters(self) -> List[np.ndarray]:
             The local model parameters as a list of NumPy ndarrays.
         """
 
+    # @abstractmethod
+    def get_properties(self, ins: PropertiesIns) -> PropertiesRes:
```
Comment (danieljanes) [inline src/py/flwr/client/numpy_client.py:93]: The argument should be called `config` and the type should be the same as in `fit/evaluate`

Comment (danieljanes) [reply src/py/flwr/client/numpy_client.py:93]: Some background: `Ins/Res` types are only used in `Client` whereas `NumPyClient` uses higher-level / more Pythonic / easier-to-use types

**Review by danieljanes — CHANGES_REQUESTED** [review]

File: src/py/flwr/client/numpy_client.py (hunk 5 above)
Comment (pedropgusmao) [reply src/py/flwr/client/numpy_client.py:93]: As in `Ins/Res` use gRPC messages, and `NumpyClient` uses dictionary?

File: src/proto/flwr/proto/transport.proto (hunk 4 above)
Comment (pedropgusmao) [reply src/proto/flwr/proto/transport.proto]: @danieljanes same for `PropertiesRes` ?

File: src/py/flwr/client/numpy_client.py [hunk 6]
```diff
@@ -175,6 +189,10 @@ class NumPyClientWrapper(Client):
 
     def __init__(self, numpy_client: NumPyClient) -> None:
         self.numpy_client = numpy_client
+        self.properties: Properties = {"tensor_type": "numpy.ndarray"}
+
+    def get_properties(self, ins: PropertiesIns) -> PropertiesRes:
+        return PropertiesRes(properties=self.properties)
```
Comment (danieljanes) [inline src/py/flwr/client/numpy_client.py:195]: Using self.properties here means that `NumPyClient.get_properties` never gets called. My recommendation is to keep `NumPyClientWrapper` stateless (apart from the `numpy_client` property). The most flexible approach would be to leave setting ` {"tensor_type": "numpy.ndarray"}` up to the user.

File: src/py/flwr/client/keras_client.py [hunk 7]
```diff
@@ -108,13 +111,17 @@ class KerasClientWrapper(Client):
...
+        self.properties: Properties = {"tensor_type": "numpy.ndarray"}
 
     def get_parameters(self) -> ParametersRes:
         """Return the current local model parameters."""
         weights = self.keras_client.get_weights()
         parameters = weights_to_parameters(weights)
         return ParametersRes(parameters=parameters)
 
+    def get_properties(self, ins: PropertiesIns) -> PropertiesRes:
+        return PropertiesRes(properties=self.properties)
```
Comment (danieljanes) [inline src/py/flwr/client/keras_client.py:123]: `KerasClientWrapper` should ideally remain stateless (apart from the `keras_client` property, see comments on `NumPyClientWrapper` for comparison)

File: src/py/flwr/client/keras_client.py [hunk 8]
```diff
@@ -108,13 +111,17 @@ class KerasClientWrapper(Client):
 
     def __init__(self, keras_client: KerasClient) -> None:
         self.keras_client = keras_client
+        self.properties: Properties = {"tensor_type": "numpy.ndarray"}
```
Comment (danieljanes) [inline src/py/flwr/client/keras_client.py:114]: This should be removed, see comment below

**Review by danieljanes — CHANGES_REQUESTED** [review]
Thanks for the updates @pedropgusmao. Just a few minor things left, almost there!

File: src/py/flwr/client/numpy_client.py (hunk 6 above)
Comment (pedropgusmao) [reply src/py/flwr/client/numpy_client.py]: I removed `self.properties` from wrapper and forwarded the `get_properties` call to `numpy_client` attribute.

**Review by danieljanes — APPROVED** [review]

**Pull request merged**